src/core/observable.ts

```diff
@@ -131,6 +132,8 @@ export function endBatch() {
 }
 
 export function reportObserved(observable: IObservable): boolean {
+    checkIfStateReadsAreAllowed(observable)
```

Comment:
NIT: I think this can be moved into the else branch at line 154? (since these lines are hit millions of time, trying to move stuff out of the critical as much as possible :))

Reply:
We need to check for reads anytime an observable is accessed regardless of `derivation`/`batch`/`observers`.
Initially the PR relied on the presence of batch/derivation, but low level tools like `untracked/transaction` were modifing the behavior, conflicting with the intention of requiring `reaction/autorun`.        
So we introduced `allowStateReads`, so we don't depend on `batch/derivation`, similary like `enforceAction` doesn't depend on a presence `batch`, but uses it's own flag `allowStateChanges`.
The way to optimize this would be to move the check into individual "reading" operations of observable structures (value/array/map/object) etc. It's discussed here: 
https://github.com/mobxjs/mobx/pull/2079#discussion_r317235819

Reply:
Check! I'm clearly late to the show :)